### PR TITLE
[visionOS] Provide support for building Firestore on visionOS

### PR DESCRIPTION
**Pull request by ncooke3**
#no-changelog

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -1348,3 +1305,170 @@ func googleAppMeasurementDependency() -> Package.Dependency {
...
+  // "firebase" prefix to prevent any API collisions (such issues should not
+  // arise as the manifest APIs should be confined to the `Package.swift`).
+  static var firebaseVisionOS: Self {
+    #if swift(>=5.9)
+      // For Xcode 15, return the available `visionOS` platform.
+      return .visionOS
+    #else
+      // For Xcode 14, return `iOS` as visionOS is unavailable. Since all targets
+      // support iOS, this acts as a no-op.
+      return .iOS
```
Comment (andrewheard) [inline Package.swift:1471]: I assume Xcode 14 doesn't complain about `.iOS` being listed twice in the platforms array?

**Review by andrewheard — APPROVED** [review]
LGTM, thanks!

File: Firestore/CHANGELOG.md [hunk 2]
```diff
@@ -3,6 +3,15 @@
...
   (e.g. while the client was offline). (#INSERT_PR_NUMBER_HERE)
+- [added] Developers using Firestore on **visionOS** must use a source
+  Firestore distribution rather than the default binary distribution. To do
+  this, quit Xcode and open the desired project from the command line
+  with the `FIREBASE_SOURCE_FIRESTORE` environment variable:
+  ```
+  open --env FIREBASE_SOURCE_FIRESTORE /path/to/project.xcodeproj
+  ```
+  To go back to using the binary distribution of Firestore, quit Xcode and
+  open Xcode like normal, without the environment variable. (#11492)
```
Comment (ncooke3) [inline Firestore/CHANGELOG.md:14]: Hi @morganchen12, WDYT of this?

Comment (morganchen12) [reply Firestore/CHANGELOG.md:14]: It's not an ideal UX, but we can improve it as visionOS adoption increases. For the time being I think it's fine.

**Review by morganchen12 — APPROVED** [review]

File: Package.swift (hunk 1 above)
Comment (ncooke3) [reply Package.swift:1471]: Yep, manually tested that opening this branch's Package.swift in 14.3 resolved successfully.

There has been some precedent for featuring a platform twice in an array: https://github.com/google/GoogleDataTransport/blob/main/Package.swift#L109

File: Package.swift [hunk 3]
```diff
@@ -1348,3 +1305,170 @@ func googleAppMeasurementDependency() -> Package.Dependency {
...
+  if ProcessInfo.processInfo.environment["FIREBASE_SOURCE_FIRESTORE"] != nil {
+    packageInfo = (
+      "https://github.com/firebase/abseil-cpp-SwiftPM.git",
+      "0.20220623.0" ..< "0.20220624.0"
+    )
+  } else {
+    packageInfo = (
+      "https://github.com/google/abseil-cpp-binary.git",
+      "1.2022062300.0" ..< "1.2022062400.0"
+    )
```
Comment (ncooke3) [inline Package.swift:1323]: Previously, the binary abseil dep looked like:
```
    .package(
      url: "https://github.com/google/abseil-cpp-binary.git",
      "1.2022062300.0" ..< "1.2022062400.0"
    ),
```
So the URL and version range remains unchanged, as intended.

File: Package.swift [hunk 4]
```diff
@@ -1348,3 +1305,170 @@ func googleAppMeasurementDependency() -> Package.Dependency {
...
+
+func grpcDependency() -> Package.Dependency {
+  let packageInfo: (url: String, range: Range<Version>)
+
+  // If building Firestore from source, abseil will need to be built as source
+  // as the headers in the binary version of abseil are unusable.
+  if ProcessInfo.processInfo.environment["FIREBASE_SOURCE_FIRESTORE"] != nil {
+    packageInfo = ("https://github.com/grpc/grpc-ios.git", "1.50.1" ..< "1.51.0")
+  } else {
+    packageInfo = ("https://github.com/google/grpc-binary.git", "1.50.1" ..< "1.51.0")
```
Comment (ncooke3) [inline Package.swift:1337]: Previously, the binary grpc dep looked like:
```
    .package(
      url: "https://github.com/google/grpc-binary.git",
      "1.50.1" ..< "1.51.0"
    ),
```
So the URL and version range remains unchanged, as intended.

File: Package.swift [hunk 5]
```diff
@@ -1348,3 +1305,170 @@ func googleAppMeasurementDependency() -> Package.Dependency {
...
+  return .package(url: packageInfo.url, packageInfo.range)
+}
+
+func grpcDependency() -> Package.Dependency {
+  let packageInfo: (url: String, range: Range<Version>)
+
+  // If building Firestore from source, abseil will need to be built as source
+  // as the headers in the binary version of abseil are unusable.
+  if ProcessInfo.processInfo.environment["FIREBASE_SOURCE_FIRESTORE"] != nil {
+    packageInfo = ("https://github.com/grpc/grpc-ios.git", "1.50.1" ..< "1.51.0")
```
Comment (ncooke3) [inline Package.swift:1335]: The source based gRPC versions match that of binary gRPC's version range.

File: Package.swift [hunk 6]
```diff
@@ -1348,3 +1305,170 @@ func googleAppMeasurementDependency() -> Package.Dependency {
...
+func abseilDependency() -> Package.Dependency {
+  let packageInfo: (url: String, range: Range<Version>)
+
+  // If building Firestore from source, abseil will need to be built as source
+  // as the headers in the binary version of abseil are unusable.
+  if ProcessInfo.processInfo.environment["FIREBASE_SOURCE_FIRESTORE"] != nil {
+    packageInfo = (
+      "https://github.com/firebase/abseil-cpp-SwiftPM.git",
+      "0.20220623.0" ..< "0.20220624.0"
+    )
```
Comment (ncooke3) [inline Package.swift:1318]: The source based abseil versions are equivalent to that of binary abseil's version range. 

These binary abseil has a major version of 1 and adds two zeros to the minor. 
_Lower bound: 0.20220623.0 → >1<.20220623>00<.0 → 1.2022062300.0_
_Upper bound: 0.20220624.0 → >1<.20220624>00<.0 → 1.2022062400.0_

**Review by ncooke3 — COMMENTED** [review]
Adding some comments of the rational behind the versioning strategy.

**Pull request merged**